### ISSUE TITLE
Support hashstyle routing

### ISSUE THREAD

**Issue by ensc** [leptos-rs/leptos#2184]
**Is your feature request related to a problem? Please describe.**

atm, laptos_router works only with the url path.  E.g. when http://localhost/app/ serves

```rust
    view! {
        <Router>
	    <a href="home">Home</a>
	    <Routes>
	    <Route path="home" view=|| { view! { Home } }/>
	    </Routes>
	    </Router>
    }
```

a click on "Home" will go to http://localhost/app/home.

This creates problems with passive hosting (e.g. as gitlab/github pages) because the url does not exist and you will get an 404 error when trying to open it directly.  

You have to modify server configuration to make it work (`RewriteEngine on`).  Even in this case, it complicates accessing relative resources because `document.baseURI` has changed and you will have to hardcode the application uri.  It pollutes also the namespace and hides invalid URLs.

**Describe the solution you'd like**

It would be nice when routing is implemented as html anchors.  E.g. the example above would go to http://localhost/app/#home

Implementing it manually as

```rust
    view! {
        <Router>
	    <a href="#home">Home</a>
	    <Routes>
	    <Route path="#home" view=|| { view! { Home } }/>
	    </Routes>
	    </Router>
    }
```

does not work (laptos-router 0.5.4).

**Additional context**

see e.g. https://router.vuejs.org/guide/essentials/history-mode.html

**Comment by gbj** [comment]
Sounds good! Would you like to work on a PR?

**Comment by zoomiti** [comment]
I can work on this.

**Comment by nag763** [comment]
Hello ! Any update on this topic ? It would be helpful when the site is hosted on GH Pages for example! 
Thanks !

**Comment by FloWi** [comment]
Hi! 
I wish this feature existed.

I have some time this week and can try implementing it if I could get some help. I'm new to leptos and only used the csr side and never really wrote macros myself (only used them), but I already like the project and would be happy to contribute. 

For spa's to be useful it should allow the same segments in the fragment like the normal path does. 

E.g. `/#/order/:order_id`


Could someone of the maintainers point me in the right direction, if that issue is still interesting for the project? 

Maybe we can start with a broken testcase. In the above example we could have it resolve to sth like 

```rust
vec![
EmptyPathSegment,
StaticFragmentSegment("order"), ParameterFragmentSegment("order_id")]
```

I just wrote what comes to mind as a start for a discussion - haven't studied the codebase deeply yet.

**Comment by gbj** [comment]
Happy to provide some pointers.

The router is designed to support multiple different "location providers," but it has only ever had one (using the browser's History API). You can find that implementation [here](https://github.com/leptos-rs/leptos/blob/main/router/src/location/history.rs).

The only place I'm aware that it's hardcoded to use that location provider is [here](https://github.com/leptos-rs/leptos/blob/main/router/src/components.rs#L86-L87) but that's easy to work around or change.

So, the steps to take here would be to implement an alternate `LocationProvider` implementation for `HashRouter` or whatever, which provides the same functionality as `BrowserUrl` (the History API implementation) but uses events list `hashchange` instead of `popstate`, etc. (There may be some places where the implementation has leaked out of the `LocationProvider` abstraction since it's only ever had a single implementation; those can be fixed as needed.)

This falls into the category of "I'm happy to have someone else work on this and offer some pointers, but am not interested in working on it myself," so -- feel free to ask questions etc.!

**Comment by FloWi** [comment]
Nice, thanks for all the pointers! Will take a look tomorrow and see how far I get. 

Totally understandable to not have that feature implemented when you don't need it, but I appreciate your support. It'd come in handy when your want to use gh-pages. The other hacks via custom 404 page feel dirty 🙈

**Comment by FloWi** [comment]
I experimented with the codebase, but I think this rabbit-hole is way too deep for me as a leptos newbie and I feel my lack of understanding of the browser navigation apis could cause a lot of trouble.

It looks like the hash stuff in the browsers is closely related to the historyAPI, and I think there would be a lot of code duplication if we decide to implement a 2nd `LocationProvider` that just listens to the hashchange events. Or maybe I'm missing something.

After giving it more thought I found out that having just a HashRouter impl wouldn't actually work for my scenario. So I decided to go a different route (:D) and host the static site somewhere else.

**Comment by kahboon0425** [comment]
> Hello ! Any update on this topic ? It would be helpful when the site is hosted on GH Pages for example! Thanks !

Hi @nag763, I found a workaround for this! You just need to copy your `index.html` to `404.html` for GitHub pages to use that when it goes 404 haha.

**Comment by mohe2015** [comment]
I looked a bit into making the `LocationProvider` configurable but as the trait is not dyn compatible it seems the obvious solution of storing it as `Box<dyn LocationProvider>` does not work. Would it make sense to separate the non-self methods from that trait? So try a larger redesign of that trait? Or do you have another idea?

**Comment by mohe2015** [comment]
Treading through the generics the most obvious way https://github.com/mohe2015/leptos/commit/34a6f8a9911d98820efd33e665286530e0c8a775 seems to not work nicely.

**Comment by mohe2015** [comment]
I experimented a lot with this but failed to implement this in a way that I felt confident with and that was working properly. I think the main problem is that many places in the code currently assume the routing url to be equal to the browser url. Also the conversions are inconsistent at some places I think and never specified explicitly. In general I found it hard to understand the code. I think as this touches quote a few core places in the router it would be better suited for a knowledgeable contributor / member to implement this.

**Comment by mohe2015** [comment]
For creating browser extensions with leptos there are unfortunately no workarounds as you can not serve files at arbitrary paths.

**Comment by kevalthumar-inventyv** [comment]
Hey @FloWi , Any update on this?

**Comment by kevalthumar-inventyv** [comment]
Hi @gbj , Do you have a plan to support this in the next release?

**Comment by gbj** [comment]
> Hi @gbj , Do you have a plan to support this in the next release?

@kevalthumar-inventyv Nope! My last comment remains true: This is not a feature I'd be interested in using, and not one that I'm going to spend time working on. If someone else is interested in working on it and submitting a PR, I'd be happy to review it.

(Note that the other person you pinged also says in their comment a year and a half ago that they decided this wasn't the right route and stopped pursuing it.)

**Comment by kevalthumar-inventyv** [comment]
Okay, thanks for your response
